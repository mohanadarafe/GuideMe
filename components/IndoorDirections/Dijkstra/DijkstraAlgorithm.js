/**
 * Implementation of Dijkstra's shortest path algorithm
 * @param {*} graph | graph of nodes (ClassGraph.js)
 * @param {*} start | starting node
 * @param {*} end | ending node
 */
export const dijkstra = (graph, start, end) => {
    // track the lowest cost to reach each node
    let costs = {};
    costs[end] = "Infinity";
    costs = Object.assign(costs, graph[start]);

    // track paths
    const parents = {end: null};
    for (let child in graph[start]) {
        parents[child] = start;
    }

    const processed = [];
    let node = lowestCostNode(costs, processed);

    while (node) {
        let cost = costs[node];
        let children = graph[node];
        for (let n in children) {
            if (String(n) !== String(start)) {
                let newCost = cost + children[n];
                if (!costs[n] || costs[n] > newCost) {
                    costs[n] = newCost;
                    parents[n] = node;
                }
            }
        }
        processed.push(node);
        node = lowestCostNode(costs, processed);
    }

    let optimalPath = [end];
    let parent = parents[end];
    while (parent) {
        optimalPath.push(parent);
        parent = parents[parent];
    }
    optimalPath.reverse();

    const results = {
        distance: costs[end],
        path: optimalPath
    };

    return results;
};

// Looks for cost of paths & goes with cheapest in cost
const lowestCostNode = (costs, processed) => {
    return Object.keys(costs).reduce((lowest, node) => {
        if (lowest === null || costs[node] < costs[lowest]) {
            if (!processed.includes(node)) {
                lowest = node;
            }
        }
        return lowest;
    }, null);
};


/**
 * Get the floor number of a specific classroom. 
 * Ex: H829 returns 8
 * @param {*} name | classroom name
 */
export const getFloorNumber = (name) => {
    if (name) {
        const num = name.match(/\d+/g).map(Number);
        
        if (name.includes("_")) {
            const room = num[0].toString();
            return room[0];
        }
        if(num.length > 1) {
            return num[0].toString();
        }
        if (num.toString().length == 3) {
            return num.toString().charAt(0)
        }
        if (num.toString().length == 4) {
            return num.toString().charAt(0) + num.toString().charAt(1);
        }   
    }
}

/**
 * Only used to fetch information from data
 * 
 * Change the classroom name to H8XX because our
 * data file stores uses associative arrays with H8XX as the key names. 
 * Ex: H529 returns H829
 * @param {*} name | classroom name
 */
export const ConvertToHall8Floor = (name) => {
    if (name) {
        const num = name.match(/\d+/g).map(Number);
        const numToString = num.toString();
        
        if(numToString.charAt(0) == 9) {
            return name;
        }
        else if (numToString.length == 3) {
            return name.replace(numToString.charAt(0), "8");
        }
        else if (numToString.length == 4) {
            return name.replace(numToString.substring(0,2), "8");
        }
    }  
}
