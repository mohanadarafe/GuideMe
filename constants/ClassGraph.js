export function ClassGraph () {
    const graph = {
        H801: {H803: 1, H863: 1},
        H803: {H805: 1, H801: 1},
        H805: {H807: 1, H803: 1},
        H807: {checkpoint1: 1, H805: 1},
        checkpoint1: {H806: 1, H811: 1, H807: 1, exit: 1},
        exit: {checkpoint1: 1},
        H806: {checkpoint4: 1, checkpoint1: 1},
        H811: {H813: 1, checkpoint1: 1}, 
        H813: {H815: 1, H811: 1},
        H815: {H817: 1, H813: 1},
        H817: {H819: 1, H815: 1},
        H819: {H821: 1, H817: 1},
        H821: {H823: 1, H886: 0.5, H819: 1},
        H823: {H825: 1, H821: 1},
        H825: {H827: 1, H823: 1},
        H827: {H828: 1, H822: 0.5, H825: 1},
        H828: {H829: 1, H827: 1, H822: 1},
        H829: {H831: 0.5, H828: 1},
        H831: {H833: 1, H829: 1},
        H833: {H835: 1, H831: 1},
        H835: {H837: 1, H833: 1},
        H837: {checkpoint2: 1, H835: 1},
        checkpoint2: {H841: 1, H838: 1, H837: 1, H832: 1},
        H841: {H843: 1, checkpoint2: 1},
        H843: {H845: 1, H841: 1},
        H845: {H847: 1, H843: 1},
        H847: {H849: 1, H845: 1},
        H849: {H851: 1, H847: 1},
        H851: {H853: 1, H849: 1},
        H853: {H851: 1, H855: 1, H852: 0.5},
        H855: {H857: 1, H853: 1, H854: 0.5},
        H857: {H859: 1, H860: 0.5, H855: 1},
        H859: {checkpoint3: 1, H861: 1, H857: 1, elevator: 2},
        H861: {H863: 1, H859: 1},
        H863: {H865: 1, H867: 1, H801: 1, H861: 1},
        checkpoint3: {H862: 1},
        H862: {checkpoint4: 2, checkpoint3: 1, elevator: 1},
        checkpoint4: {H820: 1, H862: 1, H806: 1, elevator: 1},
        H820: {checkpoint4: 1, H832: 1},
        H832: {H838: 1},
        H838: {checkpoint2: 1, H820: 1},
        elevator: {H859: 1, checkpoint4: 1}
    };
    return graph;
}
