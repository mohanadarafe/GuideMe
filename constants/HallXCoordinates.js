export function HallXCoordinates (floorNumber) {
    var rooms = [];
    rooms["elevator"] = {
        x: "361",
        y: "353",
        nearestPoint: {
            x: "362",
            y: "415"
        }
    }
    rooms["exit"] = {
        x: "",
        y: "",
        nearestPoint: {
            x: "555",
            y: "43"
        }
    }
    rooms["H801"] = {
        x: "195",
        y: "175",
        nearestPoint: {
            x: "195",
            y: "217"
        }
    }
    rooms["H803"] = {
        x: "281",
        y: "155",
        nearestPoint: {
            x: "281",
            y: "217"
        }
    }
    rooms["H805"] = {
        x: "385",
        y: "155",
        nearestPoint: {
            x: "385",
            y: "217"
        }
    },
    rooms["H807"] = {
        x: "471",
        y: "155",
        nearestPoint: {
            x: "472",
            y: "217"
        }
    },
    rooms["checkpoint1"] = {
        x: "",
        y: "",
        nearestPoint: {
            x: "550",
            y: "217"
        }
    },
    rooms["H806"] = {
        x: "493",
        y: "314",
        nearestPoint: {
            x: "553",
            y: "309"
        }
    },
    rooms["H811"] = {
        x: "654",
        y: "155",
        nearestPoint: {
            x: "654",
            y: "216"
        }
    },
    rooms["H813"] = {
        x: "736",
        y: "155",
        nearestPoint: {
            x: "736",
            y: "216"
        }
    },
    rooms["H815"] = {
        x: "840",
        y: "155",
        nearestPoint: {
            x: "840",
            y: "216"
        }
    },
    rooms["H817"] = {
        x: "891",
        y: "169",
        nearestPoint: {
            x: "840",
            y: "216"
        }
    },
    rooms["H819"] = {
        x: "894",
        y: "216",
        nearestPoint: {
            x: "840",
            y: "216"
        }
    },
    rooms["H820"] = {
        x: "627",
        y: "468",
        nearestPoint: {
            x: "553",
            y: "399"
        }
    },
    rooms["H821"] = {
        x: "899",
        y: "329",
        nearestPoint: {
            x: "848",
            y: "329"
        }
    },
    rooms["H823"] = {
        x: "899",
        y: "427",
        nearestPoint: {
            x: "848",
            y: "427"
        }
    },
    rooms["H825"] = {
        x: "899",
        y: "511",
        nearestPoint: {
            x: "850",
            y: "511"
        }
    },
    rooms["H827"] = {
        x: "899",
        y: "606",
        nearestPoint: {
            x: "850",
            y: "606"
        }
    },
    rooms["H828"] = {
        x: "899",
        y: "692",
        nearestPoint: {
            x: "850",
            y: "692"
        }
    },
    rooms["H829"] = {
        x: "901",
        y: "799",
        nearestPoint: {
            x: "850",
            y: "802"
        }
    },
    rooms["H831"] = {
        x: "891",
        y: "844",
        nearestPoint: {
            x: "850",
            y: "802"
        }
    },
    rooms["H833"] = {
        x: "833",
        y: "844",
        nearestPoint: {
            x: "850",
            y: "802"
        }
    },
    rooms["H835"] = {
        x: "737",
        y: "844",
        nearestPoint: {
            x: "737",
            y: "802"
        }
    },
    rooms["H837"] = {
        x: "648",
        y: "806",
        nearestPoint: {
            x: "646",
            y: "802"
        }
    },
    rooms["H832"] = {
        x: "627",
        y: "717",
        nearestPoint: {
            x: "559",
            y: "716"
        }
    },
    rooms["H838"] = {
        x: "627",
        y: "717",
        nearestPoint: {
            x: "559",
            y: "799"
        }
    },
    rooms["checkpoint2"] = {
        x: "",
        y: "",
        nearestPoint: {
            x: "557",
            y: "804"
        }
    },
    rooms["H841"] = {
        x: "472",
        y: "845",
        nearestPoint: {
            x: "472",
            y: "802"
        }
    },
    rooms["H843"] = {
        x: "380",
        y: "845",
        nearestPoint: {
            x: "380",
            y: "802"
        }
    },
    rooms["H845"] = {
        x: "287",
        y: "845",
        nearestPoint: {
            x: "287",
            y: "802"
        }
    },
    rooms["H847"] = {
        x: "197",
        y: "845",
        nearestPoint: {
            x: "200",
            y: "802"
        }
    },
    rooms["H849"] = {
        x: "128",
        y: "845",
        nearestPoint: {
            x: "183",
            y: "802"
        }
    },
    rooms["H851"] = {
        x: "77",
        y: "808",
        nearestPoint: {
            x: "183",
            y: "802"
        }
    },
    rooms["H853"] = {
        x: "127",
        y: "691",
        nearestPoint: {
            x: "183",
            y: "691"
        }
    },
    rooms["H855"] = {
        x: "127",
        y: "601",
        nearestPoint: {
            x: "183",
            y: "601"
        }
    },
    rooms["H857"] = {
        x: "127",
        y: "512",
        nearestPoint: {
            x: "183",
            y: "512"
        }
    },
    rooms["H859"] = {
        x: "127",
        y: "426",
        nearestPoint: {
            x: "183",
            y: "415"
        }
    },
    rooms["checkpoint3"] = {
        x: "",
        y: "",
        nearestPoint: {
            x: "183",
            y: "423"
        }
    },
    rooms["H860"] = {
        x: "241",
        y: "447",
        nearestPoint: {
            x: "237",
            y: "399"
        }
    },
    rooms["H861"] = {
        x: "127",
        y: "332",
        nearestPoint: {
            x: "183",
            y: "330"
        }
    }
    rooms["H862"] = {
        x: "379",
        y: "445",
        nearestPoint: {
            x: "381",
            y: "399"
        }
    },
    rooms["H863"] = {
        x: "127",
        y: "244",
        nearestPoint: {
            x: "183",
            y: "217"
        }
    }
    rooms["checkpoint4"] = {
        x: "",
        y: "",
        nearestPoint: {
            x: "553",
            y: "401"
        }
    }
    return rooms;
}