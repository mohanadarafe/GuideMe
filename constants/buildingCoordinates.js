export default {
    "h": {
        coordinates: [
            { latitude: 45.497167, longitude: -73.579543 },
            { latitude: 45.497713, longitude: -73.579032 },
            { latitude: 45.497378, longitude: -73.578340 },
            { latitude: 45.496830, longitude: -73.578853 },
            { latitude: 45.497167, longitude: -73.579543 }
        ],
        name: "Hall Building"
    },
    "lb": {
        coordinates: [
            { latitude: 45.496708, longitude: -73.578525 },
            { latitude: 45.497291, longitude: -73.578017 },
            { latitude: 45.496896, longitude: -73.577102 },
            { latitude: 45.496255, longitude: -73.577700 },
            { latitude: 45.496708, longitude: -73.578525 }
        ],
        name: "LB"
    },
    "gm": {
        coordinates: [
            { latitude: 45.495782, longitude: -73.579149 },
            { latitude: 45.496129, longitude: -73.578806 },
            { latitude: 45.495945, longitude: -73.578429 },
            { latitude: 45.495616, longitude: -73.578742 },
            { latitude: 45.495782, longitude: -73.579149 }
        ]
    },
    "ev": {
        coordinates: [
            { latitude: 45.495574, longitude: -73.578653 },
            { latitude: 45.495908, longitude: -73.578445 },
            { latitude: 45.495697, longitude: -73.578048 },
            { latitude: 45.496039, longitude: -73.577710 },
            { latitude: 45.495825, longitude: -73.577260 },
            { latitude: 45.495175, longitude: -73.577859 },
            { latitude: 45.495574, longitude: -73.578653 }
        ],
        name: "EV Building"
    },
    "mb": {
        coordinates: [
            { latitude: 45.495523, longitude: -73.579198 },
            { latitude: 45.495358, longitude: -73.579356 },
            { latitude: 45.495014, longitude: -73.578734 },
            { latitude: 45.495200, longitude: -73.578530 },
            { latitude: 45.495523, longitude: -73.579198 }
        ],
        name: "MB Building"
    },
    "gn": {
        coordinates: [
            { latitude: 45.493397, longitude: -73.578002 },
            { latitude: 45.494382, longitude: -73.577081 },
            { latitude: 45.493848, longitude: -73.575558 },
            { latitude: 45.492615, longitude: -73.576567 },
            { latitude: 45.493397, longitude: -73.578002 }
        ],
        name: "Grey Nuns"
    },
    "sp": {
        coordinates: [
            { latitude: 45.457464, longitude: -73.642074 },
            { latitude: 45.458325, longitude: -73.641411 },
            { latitude: 45.458165, longitude: -73.640985 },
            { latitude: 45.457528, longitude: -73.641481 },
            { latitude: 45.457198, longitude: -73.640661 },
            { latitude: 45.457002, longitude: -73.640830 },
            { latitude: 45.457464, longitude: -73.642074 }
        ]
    },
    "cj": {
        coordinates: [
            { latitude: 45.457335, longitude: -73.640715 },
            { latitude: 45.457179, longitude: -73.640400 },
            { latitude: 45.457617, longitude: -73.640051 },
            { latitude: 45.457826, longitude: -73.640483 },
            { latitude: 45.457335, longitude: -73.640715 }
        ]
    },
    "cc": {
        coordinates: [
            { latitude: 45.458521, longitude: -73.640679 },
            { latitude: 45.458377, longitude: -73.640794 },
            { latitude: 45.458076, longitude: -73.639997 },
            { latitude: 45.458221, longitude: -73.639898 },
            { latitude: 45.458521, longitude: -73.640679 }
        ]
    },
    "ad": {
        coordinates: [
            { latitude: 45.457901, longitude: -73.640113 },
            { latitude: 45.458378, longitude: -73.639742 },
            { latitude: 45.458243, longitude: -73.639460 },
            { latitude: 45.457807, longitude: -73.639814 },
            { latitude: 45.457901, longitude: -73.640113 }
        ]
    },
    "vl": {
        coordinates: [
            { latitude: 45.45859546323396, longitude:  -73.6383855342865},
            { latitude: 45.45911471817363, longitude:  -73.6378812789917},
            { latitude: 45.4592351244332, longitude: -73.63816022872925 },
            { latitude: 45.45916739594383, longitude: -73.6382246017456 },
            { latitude: 45.45949851222957, longitude: -73.63913655281067 },
            { latitude: 45.45925017519755, longitude: -73.63936185836792 },
            { latitude: 45.4591372943669, longitude: -73.6393940448761 },
            { latitude: 45.459031938721054, longitude: -73.63914728164673 },
            { latitude: 45.45912224357239, longitude: -73.63909363746643 },
            { latitude: 45.45908461656855, longitude: -73.63898634910583 },
            { latitude: 45.45912976897015, longitude: -73.63893270492554 },
            { latitude: 45.45908461656855, longitude: -73.63882541656494 },
            { latitude: 45.458828752276666, longitude: -73.63903999328613 },
            { latitude: 45.45859546323396, longitude:  -73.6383855342865}
        ]
    }
};
