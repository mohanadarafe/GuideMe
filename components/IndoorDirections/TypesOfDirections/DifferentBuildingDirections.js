import React from 'react';
import { Line, G } from 'react-native-svg';
import { ClassGraph } from '../../../constants/ClassGraph';
import { Class9Graph } from '../../../constants/ClassGraph9';
import { dijkstra, getFloorNumber, ConvertToHall8Floor } from '../Dijkstra/DijkstraAlgorithm';
import { Hall9Coordinates } from '../../../constants/Hall9Coordinates';

export function DifferentBuildingDirections(props) {
    var floorFrom;
    var floorTo;
    var rooms = props.rooms;
    var graph = ClassGraph(); 
    var goTo = props.from;
    var tempGoTo = props.from;

    if (props.from.includes(" ")) {
        floorTo = getFloorNumber(props.to);
        tempGoTo = props.to;
        goTo = floorTo == 9 ? props.to : ConvertToHall8Floor(props.to);
    }

    if (props.to.includes(" ")) {
        floorFrom = getFloorNumber(props.from);
        tempGoTo = props.from;
        goTo = floorFrom == 9 ? props.from : ConvertToHall8Floor(props.from);
    }

    const pathToElevator = dijkstra((floorFrom == 9 || floorTo == 9) ? Class9Graph() : graph, goTo, "elevator").path;
    const pathToClass = dijkstra(graph, "elevator", "exit").path;

    var getNextRoomElevator = (index) => {
        if (index < pathToElevator.length){
            return pathToElevator[++index];
        }
    }

    var getNextRoomClass = (index) => {
        if (index < pathToClass.length){
            return pathToClass[++index];
        }
    }

    const linesToExit = [];
    pathToClass.forEach((element, index) => {
        if (index < pathToClass.length - 1) {
            linesToExit.push({x1: rooms[element].nearestPoint.x, y1: rooms[element].nearestPoint.y, x2: rooms[getNextRoomClass(index)].nearestPoint.x, y2: rooms[getNextRoomClass(index)].nearestPoint.y})
        }
    });

    rooms = (floorFrom == 9 || floorTo == 9) ? Hall9Coordinates() : rooms;
    const linesToElevator = [];
    pathToElevator.forEach((element, index) => {
        if (index < pathToElevator.length - 1) {
            linesToElevator.push({x1: rooms[element].nearestPoint.x, y1: rooms[element].nearestPoint.y, x2: rooms[getNextRoomElevator(index)].nearestPoint.x, y2: rooms[getNextRoomElevator(index)].nearestPoint.y})
        }
    });

    // Go to elevator
    if(props.floor == getFloorNumber(goTo)) {
        rooms = props.floor == 9 ? Hall9Coordinates() : rooms;
        return(
            <G>
                <Line x1={rooms[goTo].x} y1={rooms[goTo].y} x2={rooms[goTo].nearestPoint.x} y2={rooms[goTo].nearestPoint.y} stroke="blue" strokeWidth="5"/>
                {
                    linesToElevator.map(el => <Line x1={el.x1} y1={el.y1} x2={el.x2} y2={el.y2} stroke="blue" strokeWidth="5"/>)
                }
                <Line x1={rooms["elevator"].x} y1={rooms["elevator"].y} x2={rooms["elevator"].nearestPoint.x} y2={rooms["elevator"].nearestPoint.y} stroke="blue" strokeWidth="5"/>
            </G>
        )
    } 
    if(props.floor == 1) {
        rooms = props.rooms;
        return(
            <G>
                <Line x1={rooms["elevator"].x} y1={rooms["elevator"].y} x2={rooms["elevator"].nearestPoint.x} y2={rooms["elevator"].nearestPoint.y} stroke="blue" strokeWidth="5"/>
                {
                    linesToExit.map(el => <Line x1={el.x1} y1={el.y1} x2={el.x2} y2={el.y2} stroke="blue" strokeWidth="5"/>)
                }
            </G>
        )
    }
    return(
        <G></G>
    )
}
