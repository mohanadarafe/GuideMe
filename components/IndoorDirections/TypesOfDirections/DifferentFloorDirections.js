import React from 'react';
import { Line, G } from 'react-native-svg';
import { ClassGraph } from '../../../constants/ClassGraph';
import { dijkstra, getFloorNumber, ConvertToHall8Floor } from '../Dijkstra/DijkstraAlgorithm';
import { Hall9Coordinates } from '../../../constants/Hall9Coordinates';
import { Class9Graph } from '../../../constants/ClassGraph9';

export function DifferentFloorDirections(props) {
    const floorFrom = getFloorNumber(props.from);
    const floorTo = getFloorNumber(props.to);
    const graph = ClassGraph();

    const pathToElevator = dijkstra(floorFrom == 9 ? Class9Graph() : graph, floorFrom == 9 ? props.from : ConvertToHall8Floor(props.from), "elevator").path;
    const pathToClass = dijkstra(floorTo == 9 ? Class9Graph() : graph, "elevator", floorTo == 9 ? props.to : ConvertToHall8Floor(props.to)).path;

    var getNextRoomElevator = (index) => {
        if (index < pathToElevator.length){
            return pathToElevator[++index];
        }
    }

    var getNextRoomClass = (index) => {
        if (index < pathToClass.length){
            return pathToClass[++index];
        }
    }

    var rooms = floorFrom == 9 ? Hall9Coordinates() : props.rooms;
    const linesToElevator = [];
    pathToElevator.forEach((element, index) => {
        if (index < pathToElevator.length - 1) {
            linesToElevator.push({x1: rooms[element].nearestPoint.x, y1: rooms[element].nearestPoint.y, x2: rooms[getNextRoomElevator(index)].nearestPoint.x, y2: rooms[getNextRoomElevator(index)].nearestPoint.y})
        }
    });

    rooms = floorTo == 9 ? Hall9Coordinates() : props.rooms;
    const linesToClass = [];
    pathToClass.forEach((element, index) => {
        if (index < pathToClass.length - 1) {
            linesToClass.push({x1: rooms[element].nearestPoint.x, y1: rooms[element].nearestPoint.y, x2: rooms[getNextRoomClass(index)].nearestPoint.x, y2: rooms[getNextRoomClass(index)].nearestPoint.y})
        }
    });
    

    // Go to elevator
    if(props.floor == getFloorNumber(props.from)) {
        rooms = floorFrom == 9 ? Hall9Coordinates() : props.rooms;
        var className = ConvertToHall8Floor(props.from.toString());
        return(
            <G>
                <Line x1={rooms[className].x} y1={rooms[className].y} x2={rooms[className].nearestPoint.x} y2={rooms[className].nearestPoint.y} stroke="blue" strokeWidth="5"/>
                {
                    linesToElevator.map(el => <Line x1={el.x1} y1={el.y1} x2={el.x2} y2={el.y2} stroke="blue" strokeWidth="5"/>)
                }
                <Line x1={rooms["elevator"].x} y1={rooms["elevator"].y} x2={rooms["elevator"].nearestPoint.x} y2={rooms["elevator"].nearestPoint.y} stroke="blue" strokeWidth="5"/>
            </G>
        )
    } 
    if(props.floor == getFloorNumber(props.to)) {
        rooms = floorTo == 9 ? Hall9Coordinates() : props.rooms;
        var className = ConvertToHall8Floor(props.to.toString());
        return(
            <G>
                <Line x1={rooms["elevator"].x} y1={rooms["elevator"].y} x2={rooms["elevator"].nearestPoint.x} y2={rooms["elevator"].nearestPoint.y} stroke="blue" strokeWidth="5"/>
                {
                    linesToClass.map(el => <Line x1={el.x1} y1={el.y1} x2={el.x2} y2={el.y2} stroke="blue" strokeWidth="5"/>)
                }
                <Line x1={rooms[className].x} y1={rooms[className].y} x2={rooms[className].nearestPoint.x} y2={rooms[className].nearestPoint.y} stroke="blue" strokeWidth="5"/>
            </G>
        )
    }
    return(
        <G></G>
    )
}
