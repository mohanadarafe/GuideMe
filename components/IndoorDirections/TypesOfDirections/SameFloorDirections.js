import React from 'react';
import { Line, G } from 'react-native-svg';
import { ClassGraph } from '../../../constants/ClassGraph';
import { Class9Graph } from '../../../constants/ClassGraph9';
import { dijkstra, getFloorNumber, ConvertToHall8Floor } from '../Dijkstra/DijkstraAlgorithm';
import { Hall9Coordinates } from '../../../constants/Hall9Coordinates';

export function SameFloorDirections(props) {
    const floor = getFloorNumber(props.from);
    const rooms = floor == 9 ? Hall9Coordinates() : props.rooms;
    const graph = floor == 9 ? Class9Graph() : ClassGraph();
    const from = floor == 9 ? props.from : ConvertToHall8Floor(props.from);
    const to = floor == 9 ? props.to : ConvertToHall8Floor(props.to);
    const route = dijkstra(graph, from, to).path;

    var getNextRoom = (index) => {
        if (index < route.length){
            return route[++index];
        }
    }

    const lines = [];
    route.forEach((element, index) => {
        if (index < route.length - 1) {
            lines.push({x1: rooms[element].nearestPoint.x, y1: rooms[element].nearestPoint.y, x2: rooms[getNextRoom(index)].nearestPoint.x, y2: rooms[getNextRoom(index)].nearestPoint.y})
        }
    });

    if(props.floor == floor) {
        return(
            <G>
                <Line x1={rooms[from].x} y1={rooms[from].y} x2={rooms[from].nearestPoint.x} y2={rooms[from].nearestPoint.y} stroke="blue" strokeWidth="5"/>
                {
                    lines.map(el => <Line x1={el.x1} y1={el.y1} x2={el.x2} y2={el.y2} stroke="blue" strokeWidth="5"/>)
                }
                <Line x1={rooms[to].x} y1={rooms[to].y} x2={rooms[to].nearestPoint.x} y2={rooms[to].nearestPoint.y} stroke="blue" strokeWidth="5"/>
            </G>
        )
    }
    return(<G></G>);
}
