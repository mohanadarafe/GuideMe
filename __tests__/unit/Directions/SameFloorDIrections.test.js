import React from "react";
import renderer from "react-test-renderer";
import { HallXCoordinates } from "../../../constants/HallXCoordinates";
import { SameFloorDirections } from "../../../components/IndoorDirections/TypesOfDirections/SameFloorDirections";
import { getFloorNumber } from "../../../components/IndoorDirections/Dijkstra/DijkstraAlgorithm";

describe("SameFloorDirections component", () => {
    test("renders correctly", () => {
        const tree = renderer.create(<SameFloorDirections floor={8} rooms={HallXCoordinates()} from={"H835"} to={"H835"}/>).toJSON();
        expect(tree).toMatchSnapshot();
    });

    test("returns correct floor from classroom", () => {
        const testCases = [
            "H829",
            "H531",
            "H123",
            "H1055",
            "H1347",
            "MB3.125",
            "MB12.246",
        ]
        const results = [
            "8",
            "5",
            "1",
            "10",
            "13",
            "3",
            "12"
        ]
        testCases.forEach((element, index) => {
            expect(getFloorNumber(element)).toEqual(results[index])
        });
    })
});
