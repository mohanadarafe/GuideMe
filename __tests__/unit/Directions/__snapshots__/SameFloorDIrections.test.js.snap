// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SameFloorDirections component renders correctly 1`] = `
<RNSVGGroup
  fill={
    Array [
      0,
      4278190080,
    ]
  }
  fillOpacity={1}
  fillRule={1}
  font={Object {}}
  matrix={
    Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ]
  }
  opacity={1}
  originX={0}
  originY={0}
  propList={Array []}
  rotation={0}
  scaleX={1}
  scaleY={1}
  skewX={0}
  skewY={0}
  stroke={null}
  strokeDasharray={null}
  strokeDashoffset={null}
  strokeLinecap={0}
  strokeLinejoin={0}
  strokeMiterlimit={4}
  strokeOpacity={1}
  strokeWidth={1}
  vectorEffect={0}
  x={0}
  y={0}
>
  <RNSVGLine
    fill={
      Array [
        0,
        4278190080,
      ]
    }
    fillOpacity={1}
    fillRule={1}
    matrix={
      Array [
        1,
        0,
        0,
        1,
        0,
        0,
      ]
    }
    opacity={1}
    originX={0}
    originY={0}
    propList={
      Array [
        "stroke",
        "strokeWidth",
      ]
    }
    rotation={0}
    scaleX={1}
    scaleY={1}
    skewX={0}
    skewY={0}
    stroke={
      Array [
        0,
        4278190335,
      ]
    }
    strokeDasharray={null}
    strokeDashoffset={null}
    strokeLinecap={0}
    strokeLinejoin={0}
    strokeMiterlimit={4}
    strokeOpacity={1}
    strokeWidth="5"
    vectorEffect={0}
    x={0}
    x1="737"
    x2="737"
    y={0}
    y1="844"
    y2="802"
  />
  <RNSVGLine
    fill={
      Array [
        0,
        4278190080,
      ]
    }
    fillOpacity={1}
    fillRule={1}
    matrix={
      Array [
        1,
        0,
        0,
        1,
        0,
        0,
      ]
    }
    opacity={1}
    originX={0}
    originY={0}
    propList={
      Array [
        "stroke",
        "strokeWidth",
      ]
    }
    rotation={0}
    scaleX={1}
    scaleY={1}
    skewX={0}
    skewY={0}
    stroke={
      Array [
        0,
        4278190335,
      ]
    }
    strokeDasharray={null}
    strokeDashoffset={null}
    strokeLinecap={0}
    strokeLinejoin={0}
    strokeMiterlimit={4}
    strokeOpacity={1}
    strokeWidth="5"
    vectorEffect={0}
    x={0}
    x1="737"
    x2="737"
    y={0}
    y1="844"
    y2="802"
  />
</RNSVGGroup>
`;
